hw/xbox/mcpx_apu.c

```diff
@@ -826,12 +849,38 @@ static void se_frame(void *opaque)
                     NV_PAVS_VOICE_PAR_STATE_ACTIVE_VOICE)) {
                 MCPX_DPRINTF("voice %d not active...!\n", d->regs[current]);
                 fe_method(d, SE2FE_IDLE_VOICE, d->regs[current]);
+            } else {
+                process_voice(d, mixbins, d->regs[current]);
             }
             MCPX_DPRINTF("next voice %d\n", d->regs[next]);
             d->regs[current] = d->regs[next];
         }
     }
 
+#if GENERATE_MIXBIN_BEEP
+    /* Inject some audio to the mixbin for debugging.
+     * Signal is 1500 Hz sine wave, phase shifted by mixbin number. */
+    for (mixbin = 0; mixbin < NUM_MIXBINS; mixbin++) {
+        for (sample = 0; sample < NUM_SAMPLES_PER_FRAME; sample++) {
+            /* Avoid multiple of 1.0 / NUM_SAMPLES_PER_FRAME for phase shift,
```

Comment:
As this has no information across frames, it can only do multiples of 1500 Hz: 3000 Hz, 4500 Hz, ...
Doing another pitch shift would break the waveform.

Also, this feature is mostly meant to test / debug the following:

* Buffer use (detect underflows / overflows)
* QEMU audio output
* Timing
* EP emulation
* Clipping
* Routing
    * VP &rarr; GP
    * GP &rarr; Scratch memory
    * Scratch memory &rarr; EP
    * EP &rarr; ACI
    * Debug output

We don't need much control over the frequency to do these things - it should just be a recognizable shape, audible, and compress well.

Once we have confirmed that VP and routing works, we can also inject samples into VP via xboxpy with much greater control. That would then be used to test things like filters and envelopes.